### PR TITLE
[data] Add `Dataset.split_proportionately` and `ray.ml.train_test_split`

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds a `Dataset.split_proportionately` method that allows the user to split a dataset using proportions. This is a very common use-case for eg. train-test splitting. The implementation is a thin wrapper over `Dataset.split_at_indices`.

Additionally, this PR adds a `ray.ml.train_test_split` function intended to provide a familiar API to ML practitioners.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
Any reason not to name this `train_test_split`? Or will that be coming in a future PR as a util function in AIR?

**Comment by amogkam** [comment]
During dogfooding, this came up as users were not able to identify that `split_at_indices` is supposed to be used for train test split. I'm worried that we will have the same problem with `split_proportionately` unless the name is obvious.

**Comment by Yard1** [comment]
There's no real advantage to having `train_test_split` as a standalone function, especially when we consider the fact that we have labels as columns inside datasets, instead of being separate objects/arrays, as in sklearn.

What we can do is make the documentation more explicit. Let me update the docstring.

**Comment by amogkam** [comment]
@Yard1 this is just a naming thing. When I see `split_proportionately` it's not obvious what it means to me. 

would also like to hear thoughts from @maxpumperla

**Comment by amogkam** [comment]
To clarify, I believe the original intention of doing this was to solve 2 problems (please correct me if I'm wrong here)

1. It's not clear to users that they should be using `split_at_indices` for train-test split, even though we had this in some of our examples.
2. Even if users are able to figure it out, being able to use `split_at_indices` for train test split requires some unnecessary boilerplate.

So this PR solves 2 partly, but 1 would still be a problem right?

Documenting the usage would help with 1, but I think renaming so that it shows up in doc search results/autocomplete would help as well.

Let me know what you think! And would like to hear opinions from other folks as well.

**Comment by richardliaw** [comment]
What's the argument for split_proportionately (or against `train_test_split`)?

The argument for `train_test_split` is that there's precedence.

**Comment by Yard1** [comment]
The argument against calling it `train_test_splot` is this is more generic, and it allows you to split it into an arbitrary number of datasets, as opposed to just 2.

**Comment by amogkam** [comment]
I feel that the primary use case for `split_proportionately` is for train test split, which we should optimize for.

The minority of use cases are covered by `split_at_indices` anyways.

**Comment by Yard1** [comment]
@clarkzinzow what do you think?

**Comment by Yard1** [comment]
I'd be happy with renaming this to `train_test_split`, but we may need to limit the functionality then, to match the new name. It would streamline the usage, though.

**Comment by Yard1** [comment]
Added a `ray.ml.train_test_split` function which calls `Dataset.split_propertionately` underneath.

File: python/ray/ml/utils/datasets.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            Defaults to False. This may be a very expensive operation with large
+            datasets.
+        seed: Fix the random seed to use for shuffle, otherwise one will be chosen
+            based on system randomness. Ignored if ``shuffle=False``.
+
+    Returns:
+        Train and test subsets as two Datasets.
+    """
+    if shuffle:
+        dataset = dataset.random_shuffle(seed=seed)
```
Comment (amogkam) [inline python/ray/ml/utils/datasets.py:40]: Is there a way to check the size of a dataset and raise a warning about this being expensive if this is too large?

File: python/ray/train/examples/train_linear_dataset_example.py [hunk 2]
```diff
@@ -20,10 +20,8 @@ def get_dataset(a, b, size) -> Dataset:
 
     dataset = get_dataset(a, b, size)
 
-    split_index = int(dataset.count() * split)
-
-    train_dataset, validation_dataset = dataset.random_shuffle().split_at_indices(
-        [split_index]
+    train_dataset, validation_dataset = dataset.random_shuffle().split_proportionately(
```
Comment (amogkam) [inline python/ray/train/examples/train_linear_dataset_example.py:23]: Use train_test_split here?

**Review by amogkam — APPROVED** [review]
A couple minor comments but looks good to me @Yard1!

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 2 above)
Comment (Yard1) [reply python/ray/train/examples/train_linear_dataset_example.py:23]: This is not an AIR example, so I didn't want to use it there

File: python/ray/ml/utils/datasets.py (hunk 1 above)
Comment (Yard1) [reply python/ray/ml/utils/datasets.py:40]: I suppose we can have some heuristic to print a warning, but imo we should add that to the random_shuffle method itself if we want to go with it. @clarkzinzow what do you think?

File: python/ray/train/examples/train_linear_dataset_example.py (hunk 2 above)
Comment (amogkam) [reply python/ray/train/examples/train_linear_dataset_example.py:23]: Ah oops- did not see that.

File: python/ray/ml/utils/datasets.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            train, test = train_test_split(ds, test_size=0.25)
+            print(train.take())  # [0, 1, 2, 3, 4, 5]
+            print(test.take())  # [6, 7]
+
+    Args:
+        dataset: Dataset to split.
+        test_size: If float, should be between 0.0 and 1.0 and represent the proportion
+            of the dataset to include in the test split. If int, represents the
+            absolute number of test samples.
+        shuffle: Whether or not to global shuffle the dataset before splitting.
```
Comment (clarkzinzow) [inline python/ray/ml/utils/datasets.py:30]:
```suggestion
        shuffle: Whether or not to globally shuffle the dataset before splitting.
```

File: python/ray/ml/utils/datasets.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        seed: Fix the random seed to use for shuffle, otherwise one will be chosen
+            based on system randomness. Ignored if ``shuffle=False``.
+
+    Returns:
+        Train and test subsets as two Datasets.
+    """
+    if shuffle:
+        dataset = dataset.random_shuffle(seed=seed)
+
+    dataset_length = dataset.count()
```
Comment (clarkzinzow) [inline python/ray/ml/utils/datasets.py:42]: Can we move this `dataset.count()` call into the `else` block below? This can be expensive if the number of rows for each block isn't available in the block metadata, so we should only calculate this if required.

File: python/ray/ml/utils/datasets.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+from typing import Optional, Tuple, Union
+from ray.data import Dataset
+
+
+def train_test_split(
+    dataset: Dataset,
+    test_size: Union[int, float],
```
Comment (clarkzinzow) [inline python/ray/ml/utils/datasets.py:7]: I noticed that this deviates from the sklearn `train_test_split` [API](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) in that there's no `train_size` arg, any particular reason for that? My guess is that having both is redundant unless you want to explicitly support subsampling (e.g. `train_test_split(test_size = 0.1, train_size=0.4)`), since `train_size` can always be the complement.

File: python/ray/ml/utils/datasets.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            ds = ray.data.range(8)
+            train, test = train_test_split(ds, test_size=0.25)
+            print(train.take())  # [0, 1, 2, 3, 4, 5]
+            print(test.take())  # [6, 7]
+
+    Args:
+        dataset: Dataset to split.
+        test_size: If float, should be between 0.0 and 1.0 and represent the proportion
+            of the dataset to include in the test split. If int, represents the
+            absolute number of test samples.
```
Comment (clarkzinzow) [inline python/ray/ml/utils/datasets.py:29]: Could you add a note that the train split will always be the complement of the test split? I know that it should be obvious, but shouldn't hurt for those that would like to see that explicitly called out.

**Review by clarkzinzow — APPROVED** [review]
LGTM overall, just some small suggestions/nits!

File: python/ray/ml/utils/datasets.py (hunk 5 above)
Comment (Yard1) [reply python/ray/ml/utils/datasets.py:7]: I decided to remove it for simplicity, as the validation code can get very extensive with both arguments. We can add it back in, but that would make the function a lot more complicated.

Comment (clarkzinzow) [reply python/ray/ml/utils/datasets.py:7]: Yeah understood, that sounds good!
